File: matlab/tsp_dynamic.m
%% TSP (travelling salesman problem)
% solved using dp - try every single option but use memory (BACKCWARDS)
% formula: g(i,s) = min_{k in S} ( C_{ik} +g(k,S-{k}) )

clc; clear; close all;

%% INPUT GRAPH HERE - DISTANCE MATRIX + STARTING NODE %%%%%%%%%%%%%%%%%%%%%
% example matrix    
%nodes:  1    2    3    4    5
graph = [nan, 10,  8,   9,   7;    %1
         10,  nan, 10,  5,   6;    %2
         8,   10,  nan, 8,   9;    %3
         9,   5,   8,   nan, 6;    %4
         7,   6,   9,   6,   nan]; %5
     
starting = 1; % starting node
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

nodes = 1:length(graph); % array of nodes

unvisited = setdiff(nodes,starting); % set of nodes left to visit

dist = nan(length(nodes), length(nodes)); % distances to store in mem
prev = nan(length(nodes), length(nodes)); % previous best nodes

prev(:,1) = starting; % all final steps point to starting

% for loop iterating through each decision (each level in the decision tree)
for i = 1:size(dist,2)
   if i==1
      for loop = unvisited
          dist(loop,i) = graph(loop,starting);
      end
   elseif i==size(dist,2)
       dist(:,i) =dist(:,i-1)+graph(starting,:)';
       min_dist = min(dist(:,i));
       idx = find(dist(:,i)==min_dist);
       route = starting;
       next = prev(idx,i-1);
       for j = i-2:-1:1
           route = [route, next];
           next = prev(next,j);
       end
       route = [route, next];
       fprintf('Best Route: '); fprintf(repmat('%d ',1,length(nodes)+1),route);
       fprintf('\nDistance: %d\n',min_dist);
   else
      for loop = unvisited
          min_dist = inf;
          idx = nan;
          for j = setdiff(unvisited,loop)
              temp = dist(j,i-1) + graph(loop,j);
              visited = [];
              for k = i-1:-1:1
                  visited = [visited, prev(j,k)]; 
              end
              if sum(ismember(visited,loop))==0 && temp < min_dist
                  min_dist = temp;
                  idx = j;
              end
          end
          dist(loop,i) = min_dist;
          prev(loop,i) = idx;
      end
   end
end
      
             

File: python/tsp_dynamic.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jul 27 02:45:32 2019

@author: Aditya
"""

# -*- coding: utf-8 -*-
"""
Created on Fri Jul 26 02:56:40 2019

@author: Aditya
"""
import numpy as np

def getMin(nums):
    minval = np.inf
    idx = np.nan
    for i in range(0,len(nums)):
        if nums[i]<minval:
            minval = nums[i]
            idx = i
    return [minval,idx]

# CHANGE THESE - - - - - - - - - - - - - - - -
# distance matrix
graph = np.array([[np.nan, 10, 8, 9, 7],
                 [10, np.nan, 10, 5, 6],
                 [8, 10, np.nan, 8, 9],
                 [9, 5, 8, np.nan, 6],
                 [7, 6, 9, 6, np.nan]])
# starting node
starting = 0  
# - - - - - - - - - - - - - - - - - - - - - - -

# list of nodes
nodes = list(range(0,len(graph[0])))

# initialize distance matrix
dist = np.empty((len(graph[0]),len(graph[0])))
dist[:] = np.nan

# initialize previous matrix
prev = np.empty((len(graph[0]),len(graph[0])))
prev[:] = np.nan
prev[:,0] = 0

# define starting point
visited = [starting]

# define unvisited nodes 
unvisited = [x for x in nodes if x not in visited]

for i in nodes:
    if i==0:
        for loop in unvisited:
            dist[loop,i] = graph[loop,starting]
    elif i==nodes[-1]:
        dist[:,i] = dist[:,i-1] + np.transpose(graph[starting,:])
        [minval,idx] = getMin(dist[:,i])
        route = [starting,idx]
        nextnode = int(prev[idx,i-1]) 
        for j in list(range(i-2,0,-1)):
            route = route + [nextnode]
            nextnode = int(prev[nextnode,j])
        route = route + [nextnode]
        print(route)
        #printRoute(route)
        print(minval)
    else:
        for loop in unvisited:
            min_dist = np.inf
            idx = np.nan
            vec = [x for x in unvisited if x not in [loop]]
            for j in vec:
                temp = dist[j,i-1] + graph[loop,j]
                visited = []
                for k in list(range(i-1,0,-1)):
                    visited = visited + [prev[j,k]]
                if loop not in visited and temp < min_dist:
                    min_dist = temp
                    idx = j
            dist[loop,i] = min_dist
            prev[loop,i] = idx
                
